### ISSUE TITLE
[DataGrid] cannot re-enter table body with keyboard after page change

### ISSUE THREAD

**Issue by icf-mrankin** [mui/mui-x#11629]
### Search keywords

datagrid keyboard accessibility pagination

### Latest version

- [X] I have tested the latest version

### Steps to reproduce

Link to live example: [mui docs](https://mui.com/x/react-data-grid/#mit-version-free-forever)

Steps:
1. visit the mui docs URL listed above
2. tab into the body of the table
3. user arrow keys to move inside the table
4. tab to the next page button and click
5. try to tab back to the body of the table

### Current behavior

Using only keyboard navigation, tab into the body of a table.  Then, use the arrow keys to move to a new location within the body. Next, click on the pagination buttons and change the page.  Try to move back into the body of the table.  The body of the table is skipped, and the focus moves to the toolbar. 

If you go back to the original page, the body can be accessed as expected.  

### Expected behavior

You should be able to access the body of the table regardless of which page is current.

### Context

This is purely an accessibility issue.  You should be able to navigate the table with the keyboard.

### Your environment

  System:
    OS: macOS 14.1
  Binaries:
    Node: 21.4.0 - /usr/local/bin/node
    Yarn: 1.22.17 - /usr/local/bin/yarn
    npm: 10.2.4 - /usr/local/bin/npm
  Browsers:
    Chrome: 120.0.6099.199
    Edge: Not Found
    Safari: 17.1
  npmPackages:
    @emotion/react: ^11.11.1 => 11.11.1 
    @emotion/styled: ^11.11.0 => 11.11.0 
    @mui/base:  5.0.0-beta.27 
    @mui/core-downloads-tracker:  5.15.0 
    @mui/icons-material: ^5.15.0 => 5.15.0 
    @mui/lab: ^5.0.0-alpha.116 => 5.0.0-alpha.156 
    @mui/material: ^5.15.0 => 5.15.0 
    @mui/private-theming:  5.15.0 
    @mui/styled-engine:  5.15.0 
    @mui/system:  5.15.0 
    @mui/types:  7.2.11 
    @mui/utils: ^5.15.0 => 5.15.0 
    @mui/x-data-grid: ^5.17.26 => 5.17.26 
    @mui/x-data-grid-pro: ^6.18.4 => 6.18.7 
    @mui/x-date-pickers: ^6.18.4 => 6.18.4 
    @mui/x-license-pro:  6.10.2 
    @types/react: ^18.2.45 => 18.2.45 
    react: ^18.2.0 => 18.2.0 
    react-dom: ^18.2.0 => 18.2.0 
    typescript:  5.2.2

**Comment by michelengelen** [comment]
Hey @icf-mrankin ... It does work for me. Is the behavior I did correct, or did you mean something else?


https://github.com/mui/mui-x/assets/32863416/ddd11dfb-300a-41e5-b2ec-b56bb253b111

**Comment by icf-mrankin** [comment]
@michelengelen very strange.  I can do pretty much the same movement and it always has issues:

https://github.com/mui/mui-x/assets/47259753/e2b527ca-617c-4bce-b7d2-4db4e6a25d52

Note at the end that I can't move back into the body.

**Comment by michelengelen** [comment]
> Note at the end that I can't move back into the body.

It is indeed strange. What browser are you using for this? I see that you are on Mac as I am, so that's not likely to be the problem

**Comment by icf-mrankin** [comment]
I see pretty much the same issue with Safari 17, Brave 1.16.116, Google Chrome 120.0.6099.199, and FireFox 121.0.1.  I don't have any proof of this, but it feels like the current focus is either hidden or the target doesn't exist after a page change.  Once you move back to the original page, the focus becomes available in the cell where it was last.

I see that the tabIndex is being set to -1 when a cell or header does not have the focus.  Could it be set up that the tabIndex on the first column header is always set to 0?

**Comment by michelengelen** [comment]
I am not able to reproduce. 
@DanailH could you give this a shot?

I tested this with Safari, Chrome and FF ... no problem to focus.
Also tested this with larger data sets and in a custom app.

**Comment by michelengelen** [comment]
@MBilalShafi could you try to replicate this?

**Comment by icf-mrankin** [comment]
One thing I've found since we started looking at this is that it only appears to be a problem with keyboard-only navigation.  If you're running a screen reader, it's not an issue.  When using voice-over or NVDA, moving the focus inside the table and then changing the page and tabbing back to the table body puts you on the first column header. 
That is what I think the behavior for keyboard navigation should be, but anything that lets me back in the body on a different page would work.

**Comment by MBilalShafi** [comment]
I can reproduce the issue on both Chrome and Firefox by the steps mentioned by @icf-mrankin, the important step is clicking the next page icon with a mouse pointer rather than the keyboard.

https://github.com/mui/mui-x/assets/12609561/727d64c4-a600-48b6-a43d-e637dab75524

I didn't dig deep but I suspect that on clicking the button with the mouse the `tabIndex` state under focus management is getting reset for some reason, due to which the control never goes back to the Grid.

**Comment by icf-mrankin** [comment]
I'm glad that somebody else can actually see the effect, lol.  In the movie I posted above, I never clicked the page navigation buttons, but I clicked the "reset focus" button.  I thought that maybe it was any mouse activity that would trigger it, so I tried isolating it in [code sandbox](https://codesandbox.io/p/sandbox/xenodochial-lederberg-vk2mh7).  

It's a little different, but I can eventually get the table body to be skipped with a combination of arrowing through the table and tabbing to the pagination buttons and activating them with either the return key or the spacebar.

**Comment by bgerm** [comment]
We are using MUI DataGrid Pro and we had the Data Grid fail our accessibility audit.  We use `paginationMode="server"`.

The user cannot tab back to the table after using the keyboard to select a data grid cell and then using the keyboard to tab to and submit the next page button.

These steps can be reproduced using the MUI [ServerPaginationGrid demo](https://codesandbox.io/embed/lwhptm?module=/src/Demo.tsx&fontsize=12).
